installers/binstall/src/system/unix.rs

```diff
@@ -0,0 +1,212 @@
+        }
+    }
+
+    fn source_string(&self, installer: &Installer) -> Result<String, InstallerError> {
+        Ok(format!(
+            r#"source "{}/env""#,
+            installer.get_base_dir_path()?.to_string_lossy()
+        ))
+    }
+}
+
+#[derive(Debug)]
+struct Posix;
+
+impl UnixShell for Posix {
+    fn does_exist(&self) -> bool {
+        true
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        match crate::get_home_dir_path().ok() {
+            Some(dir) => vec![dir.join(".profile")],
+            _ => vec![],
+        }
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        // Write to .profile even if it doesn't exist. It's the only rc in the
+        // POSIX spec so it should always be set up.
+        self.rcfiles()
+    }
+}
+
+#[derive(Debug)]
+struct Bash;
+
+impl UnixShell for Bash {
+    fn does_exist(&self) -> bool {
+        !self.update_rcs().is_empty()
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        // Bash also may read .profile, however we already includes handling
+        // .profile as part of POSIX and always does setup for POSIX shells.
+        [".bash_profile", ".bash_login", ".bashrc"]
+            .iter()
+            .filter_map(|rc| crate::get_home_dir_path().ok().map(|dir| dir.join(rc)))
+            .collect()
+    }
+
+    fn update_rcs(&self) -> Vec<PathBuf> {
+        self.rcfiles()
+            .into_iter()
+            .filter(|rc| rc.is_file())
+            .collect()
+    }
+}
+
+#[derive(Debug)]
+struct Zsh;
+
+impl Zsh {
+    fn zdotdir() -> Result<PathBuf, InstallerError> {
+        use std::ffi::OsStr;
+        use std::os::unix::ffi::OsStrExt;
+
+        if matches!(env::var("SHELL"), Ok(sh) if sh.contains("zsh")) {
+            match env::var("ZDOTDIR") {
+                Ok(dir) if !dir.is_empty() => Ok(PathBuf::from(dir)),
+                _ => Err(InstallerError::ZshSetup),
+            }
+        } else {
+            match std::process::Command::new("zsh")
+                .args(&["-c", "'echo $ZDOTDIR'"])
+                .output()
+            {
+                Ok(io) if !io.stdout.is_empty() => Ok(PathBuf::from(OsStr::from_bytes(&io.stdout))),
+                _ => Err(InstallerError::ZshSetup),
+            }
+        }
+    }
+}
+
+fn has_cmd(cmd: &str) -> bool {
+    let cmd = format!("{}{}", cmd, env::consts::EXE_SUFFIX);
+    let path = env::var_os("PATH").unwrap_or_default();
+    env::split_paths(&path)
+        .map(|p| p.join(&cmd))
+        .any(|p| p.exists())
+}
+
+impl UnixShell for Zsh {
+    fn does_exist(&self) -> bool {
+        // zsh has to either be the shell or be callable for zsh setup.
+        matches!(env::var("SHELL"), Ok(sh) if sh.contains("zsh")) || matches!(has_cmd("zsh"), true)
+    }
+
+    fn rcfiles(&self) -> Vec<PathBuf> {
+        [Zsh::zdotdir().ok(), crate::get_home_dir_path().ok()]
+            .iter()
+            .filter_map(|dir| dir.as_ref().map(|p| p.join(".zshenv")))
```

Comment:
what about `.zshrc`?

Reply:
adding to `.zshenv` allows us to add our `source $HOME/.rover/env` without making the end user reload their terminal, so there is no need to modify `.zshrc` (this is what rustup does)

Reply:
Ah, fwiw it did make me reload my terminal when using zsh sadly :(